### ISSUE TITLE
python tracing can't have symbols with `os._exit(0)`

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#1685]
When the target python has `os._exit(0)` inside, uftrace can't generate symbol tables so the tracing result looks bad.

If `p-abc.py` has `os._exit(0)` at the end as follows.
```py
$ cat p-abc.py 
#!/usr/bin/env python3

import os

def c():
    return os.getpid() % 100000

def b():
    return c() + 1

def a():
    return b() - 1

a()
os._exit(0)
```
The tracing result looks as follows.
```
$ uftrace p-abc.py 
# DURATION     TID     FUNCTION
            [ 38580] | <1>() {
            [ 38580] |   <2>() {
            [ 38580] |     <3>() {
            [ 38580] |       <4>() {
   1.350 us [ 38580] |         <5>();
   8.483 us [ 38580] |       } /* <4> */
  11.374 us [ 38580] |     } /* <3> */
  14.137 us [ 38580] |   } /* <2> */
            [ 38580] |   /* linux:task-exit */

uftrace stopped tracing with remaining functions
================================================
task: 38580
[0] <1>
```

**Comment by honggyukim** [comment]
It looks `os._exit`doesn't call destructor so can't write the symbol table at the end.

> Exit the process with status n, without calling cleanup handlers, flushing stdio buffers, etc. 

https://docs.python.org/3/library/os.html#os._exit

It would be helpful if we can override `os._exit` but not sure if it's possible.

**Comment by honggyukim** [comment]
Maybe with the same reason, if a user press `Ctrl+C` then uftrace doesn't have a chance to write symbols.

**Comment by honggyukim** [comment]
We have to do something like in `uftrace.py`.
```py
os_exit = os._exit

def on_exit(n):
    uftrace_python.dtor()    # calls uftrace_trace_python_finish() in trace-python.c
    os_exit(n)

os._exit = on_exit
```
This might also have to be done in C level inside `trace-python.c` directly.

**Comment by namhyung** [comment]
Hmm.. that means we need to intercept every possible path to the exit.  I'm not sure we can do that.

**Comment by honggyukim** [comment]
After install `mypy`, it shows an error in normal execution as follows.
```
$ python3 -m pip install mypy
Defaulting to user installation because normal site-packages is not writeable
Collecting mypy
  Downloading mypy-1.2.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 4.9 MB/s eta 
         ...
$ mypy misc/gen-autoargs.py 
misc/gen-autoargs.py:311: error: Cannot determine type of "enum_format"  [has-type]
Found 1 error in 1 file (checked 1 source file)
```
If it's recorded with uftrace, then it also shows the same error and I think the execution looks fine.
```
$ uftrace record mypy misc/gen-autoargs.py 
misc/gen-autoargs.py:311: error: Cannot determine type of "enum_format"  [has-type]
Found 1 error in 1 file (checked 1 source file)

$ uftrace replay
# DURATION     TID     FUNCTION
            [  5510] | <1>() {
   4.747 ms [  5510] |   <2>();
 164.490 ms [  5510] |   <2>();
 426.784 us [  5510] |   <4a24>();

uftrace stopped tracing with remaining functions
================================================
task: 5510
[0] <1>
```
But the replay output doesn't show symbol properly. I guess this is the same problem with this issue.

**Comment by namhyung** [comment]
Yeah, probably.  But I don't know how `mypy` works exactly.

**Comment by yihong0618** [comment]
@honggyukim  @namhyung I found a tricky to fix this maybe


just add one code in `uftrace.py`
```py
code = open(sys.argv[0]).read()
sys.setprofile(uftrace_python.trace)
code = code.replace("os._exit", "exit"). # added these
exec(code, new_globals)
sys.setprofile(None)
```
that can fix both example problem and mypy problem
and do not need to intercept every possible path to the exit.

**Comment by yihong0618** [comment]
![image](https://github.com/namhyung/uftrace/assets/15976103/2216a451-cb30-4d5f-81c5-b9e389bebe78)

**Comment by yihong0618** [comment]
> Yeah, probably. But I don't know how `mypy` works exactly.

mypy use fast_exit when done

![image](https://github.com/namhyung/uftrace/assets/15976103/57247ab9-4f51-4235-b4a6-f089f5c51a28)

**Comment by honggyukim** [comment]
Great! Thanks a lot for the analysis and fix. Could you please send a PR for the fix?

**Comment by yihong0618** [comment]
of course! and thank you very much

**Comment by yihong0618** [comment]
Done

**Comment by honggyukim** [comment]
I've wrote a temporary incomplete fix as follows.
```diff
diff --git a/python/trace-python.c b/python/trace-python.c
index 8281d36b..b882264e 100644
--- a/python/trace-python.c
+++ b/python/trace-python.c
@@ -151,9 +151,18 @@ static void (*cygprof_exit)(unsigned long child, unsigned long parent);
 /* main trace function to be called from python interpreter */
 static PyObject *uftrace_trace_python(PyObject *self, PyObject *args);
 
+static void __attribute__((destructor)) uftrace_trace_python_finish(void);
+static PyObject *uftrace_trace_python_dtor(PyObject *self, PyObject *args)
+{
+       uftrace_trace_python_finish();
+       return NULL;
+}
+
 static __attribute__((used)) PyMethodDef uftrace_py_methods[] = {
        { "trace", uftrace_trace_python, METH_VARARGS,
          PyDoc_STR("trace python function with uftrace.") },
+       { "dtor", uftrace_trace_python_dtor, METH_VARARGS,
+         PyDoc_STR("finish trace python function.") },
        { NULL, NULL, 0, NULL },
 };
 
diff --git a/python/uftrace.py b/python/uftrace.py
index c00fba22..348dbd75 100644
--- a/python/uftrace.py
+++ b/python/uftrace.py
@@ -25,6 +25,14 @@ else:
 # UFTRACE_PYMAIN must be set before importing uftrace_python
 import uftrace_python
 
+real_os_exit = os._exit
+
+def os_exit(n):
+    uftrace_python.dtor();
+    real_os_exit(n)
+
+os._exit = os_exit
+
 new_globals = globals()
 new_globals["__file__"] = pathname
 sys.path.insert(0, os.path.dirname(pathname))
```
For the above [`p-abc.py`](https://github.com/namhyung/uftrace/issues/1685#issue-1689818627) example, it gets segfault, but replay output looks fine.
```
$ uftrace record p-abc.py 
WARN: child terminated by signal: 11: Segmentation fault

$ uftrace replay
# DURATION     TID     FUNCTION
            [236167] | __main__.<module>() {
            [236167] |   a() {
            [236167] |     b() {
            [236167] |       c() {
   1.150 us [236167] |         posix.getpid();
   5.691 us [236167] |       } /* c */
   8.660 us [236167] |     } /* b */
  11.642 us [236167] |   } /* a */
            [236167] |   os_exit() {
  81.559 us [236167] |     uftrace_python.dtor();
  84.867 us [236167] |   } /* os_exit */
 105.622 us [236167] | } /* __main__.<module> */
```
Maybe we can try to fix it based on this change.

**Comment by yihong0618** [comment]
does mypy works fine either？

**Comment by honggyukim** [comment]
I can't say that it works fine on mypy example because of the segfault I showed above. But it shows the function names as expected.
```
$ uftrace record mypy misc/gen-autoargs.py
misc/gen-autoargs.py:311: error: Cannot determine type of "enum_format"  [has-type]
Found 1 error in 1 file (checked 1 source file)
WARN: child terminated by signal: 11: Segmentation fault

$ uftrace replay
# DURATION     TID     FUNCTION
            [  5035] | __main__.<module>() {
 109.025 ms [  5035] |   importlib._bootstrap._find_and_load();
 480.077 us [  5035] |   re.sub();
            [  5035] |   mypy.__main__.console_entry() {
   3.309 ms [  5035] |     os_exit();
   1.305  s [  5035] |   } /* mypy.__main__.console_entry */
   1.415  s [  5035] | } /* __main__.<module> */
```

**Comment by yihong0618** [comment]
> I can't say that it works fine on mypy example because of the segfault I showed above. But the above temp change shows the function names as expected.
> 
> ```
> 
> $ uftrace record mypy misc/gen-autoargs.py
> 
> misc/gen-autoargs.py:311: error: Cannot determine type of "enum_format"  [has-type]
> 
> Found 1 error in 1 file (checked 1 source file)
> 
> WARN: child terminated by signal: 11: Segmentation fault
> 
> 
> 
> $ uftrace replay
> 
> # DURATION     TID     FUNCTION
> 
>             [  5035] | __main__.<module>() {
> 
>  109.025 ms [  5035] |   importlib._bootstrap._find_and_load();
> 
>  480.077 us [  5035] |   re.sub();
> 
>             [  5035] |   mypy.__main__.console_entry() {
> 
>    3.309 ms [  5035] |     os_exit();
> 
>    1.305  s [  5035] |   } /* mypy.__main__.console_entry */
> 
>    1.415  s [  5035] | } /* __main__.<module> */
> 
> ```
> 
> This is tested in aarch64 fedora machine.

thanks

**Comment by honggyukim** [comment]
This solution looks much better except for that I have to find a way to pass the exit status number correctly.
```diff
diff --git a/python/trace-python.c b/python/trace-python.c
index 8281d36b..bc2f9857 100644
--- a/python/trace-python.c
+++ b/python/trace-python.c
@@ -151,9 +151,18 @@ static void (*cygprof_exit)(unsigned long child, unsigned long parent);
 /* main trace function to be called from python interpreter */
 static PyObject *uftrace_trace_python(PyObject *self, PyObject *args);

+static void __attribute__((destructor)) uftrace_trace_python_finish(void);
+static PyObject *uftrace_trace_python_dtor(PyObject *self, PyObject *args)
+{
+       uftrace_trace_python_finish();
+       _exit(0);
+}
+
 static __attribute__((used)) PyMethodDef uftrace_py_methods[] = {
        { "trace", uftrace_trace_python, METH_VARARGS,
          PyDoc_STR("trace python function with uftrace.") },
+       { "dtor", uftrace_trace_python_dtor, METH_VARARGS,
+         PyDoc_STR("finish trace python function.") },
        { NULL, NULL, 0, NULL },
 };

diff --git a/python/uftrace.py b/python/uftrace.py
index c00fba22..6dc53d8e 100644
--- a/python/uftrace.py
+++ b/python/uftrace.py
@@ -25,6 +25,11 @@ else:
 # UFTRACE_PYMAIN must be set before importing uftrace_python
 import uftrace_python

+def os_exit(n):
+    uftrace_python.dtor();
+
+os._exit = os_exit
+
 new_globals = globals()
 new_globals["__file__"] = pathname
 sys.path.insert(0, os.path.dirname(pathname))
```
It makes `p-abc.py` test work fine.
```
$ uftrace p-abc.py
# DURATION     TID     FUNCTION
            [  8974] | __main__.<module>() {
            [  8974] |   a() {
            [  8974] |     b() {
            [  8974] |       c() {
   0.667 us [  8974] |         posix.getpid();
   6.750 us [  8974] |       } /* c */
   8.708 us [  8974] |     } /* b */
  10.541 us [  8974] |   } /* a */
            [  8974] |   /* linux:task-exit */

uftrace stopped tracing with remaining functions
================================================
task: 8974
[0] __main__.<module>
```

**Comment by honggyukim** [comment]
It makes `mypy` test run fine without segfault except for the exit number.
```
$ uftrace record mypy misc/gen-autoargs.py
misc/gen-autoargs.py:311: error: Cannot determine type of "enum_format"  [has-type]
Found 1 error in 1 file (checked 1 source file)

$ uftrace replay
# DURATION     TID     FUNCTION
            [  8997] | __main__.<module>() {
 115.922 ms [  8997] |   importlib._bootstrap._find_and_load();
 484.786 us [  8997] |   re.sub();
            [  8997] |   /* linux:task-exit */

uftrace stopped tracing with remaining functions
================================================
task: 8997
[0] __main__.<module>
```

**Comment by honggyukim** [comment]
I found a way to pass the exit status number and pushed the fix at #1725.
